File: Find-if-Path-Exists-in-Graph/FiPEiG.rs
use std::collections::{HashMap, HashSet};

pub fn valid_path(n: i32, edges: Vec<Vec<i32>>, source: i32, destination: i32) -> bool {
    let mut table = HashMap::new();
    edges.into_iter().for_each(|v| {
        table.entry(v[0]).or_insert(vec![]).push(v[1].clone());
        table.entry(v[1]).or_insert(vec![]).push(v[0]);
    });
    let mut paths = HashSet::new();
    let mut cache = HashMap::new();
    helper(
        &table,
        &mut cache,
        &mut paths,
        &source,
        &destination,
        n as usize,
    )
}

fn helper(
    table: &HashMap<i32, Vec<i32>>,
    cache: &mut HashMap<i32, bool>,
    paths: &mut HashSet<i32>,
    this: &i32,
    end: &i32,
    length: usize,
) -> bool {
    if cache.get(this).is_some() {
        return false;
    }

    paths.insert(*this);
    if this == end {
        true
    } else {
        table
            .get(this)
            .unwrap_or(&vec![])
            .iter()
            .find(|v| {
                if *v == end {
                    true
                } else {
                    if length == 0 {
                        paths.remove(this);
                        false
                    } else if paths.contains(v) {
                        paths.remove(this);
                        false
                    } else {
                        let x = helper(table, cache, paths, v, end, length - 1);
                        cache.insert(*this, x);
                        x
                    }
                }
            })
            .is_some()
    }
}

fn main() {
    assert!(valid_path(
        3,
        vec![vec![0, 1], vec![1, 2], vec![2, 0]],
        0,
        2
    ));

    assert!(!valid_path(
        6,
        vec![vec![0, 1], vec![0, 2], vec![3, 5], vec![5, 4], vec![4, 3]],
        0,
        5
    ));

    assert!(valid_path(1, vec![], 0, 0));

    assert!(!valid_path(
        50,
        vec![
            vec![31, 5],
            vec![10, 46],
            vec![19, 31],
            vec![5, 1],
            vec![31, 28],
            vec![28, 29],
            vec![8, 26],
            vec![13, 23],
            vec![16, 34],
            vec![30, 1],
            vec![16, 18],
            vec![33, 46],
            vec![27, 35],
            vec![2, 25],
            vec![49, 33],
            vec![44, 19],
            vec![22, 26],
            vec![30, 13],
            vec![27, 12],
            vec![8, 16],
            vec![42, 13],
            vec![18, 3],
            vec![21, 20],
            vec![2, 17],
            vec![5, 48],
            vec![41, 37],
            vec![39, 37],
            vec![2, 11],
            vec![20, 26],
            vec![19, 43],
            vec![45, 7],
            vec![0, 21],
            vec![44, 23],
            vec![2, 39],
            vec![27, 36],
            vec![41, 48],
            vec![17, 42],
            vec![40, 32],
            vec![2, 28],
            vec![35, 38],
            vec![3, 9],
            vec![41, 30],
            vec![5, 11],
            vec![24, 22],
            vec![39, 5],
            vec![40, 31],
            vec![18, 35],
            vec![23, 39],
            vec![20, 24],
            vec![45, 12]
        ],
        29,
        46
    ));
}

File: Find-if-Path-Exists-in-Graph/FiPEiG.ml
module NodeSet = Set.Make (Int)

let eadges_in_table table edges =
  List.iter
    (fun v ->
      let a = List.nth v 0 and b = List.nth v 1 in
      (match Hashtbl.find_opt table a with
      | Some l -> Hashtbl.replace table a (b :: l)
      | None -> Hashtbl.add table a [ b ]);
      match Hashtbl.find_opt table b with
      | Some l -> Hashtbl.replace table b (a :: l)
      | None -> Hashtbl.add table b [ a ])
    edges

let valid_path steps edges source dest =
  let rec helper (table : (int, int list) Hashtbl.t)
      (cache : (int, bool) Hashtbl.t) (paths : NodeSet.t ref) this endd length =
    if Hashtbl.mem cache this then false
    else if this = endd then true
    else
      (paths := NodeSet.add this !paths;
       List.find_opt
         (fun v ->
           if v = endd then true
           else if length = 0 then (
             paths := NodeSet.remove this !paths;
             false)
           else if NodeSet.mem v !paths then (
             paths := NodeSet.remove this !paths;
             false)
           else
             let x = helper table cache paths v endd (length - 1) in
             Hashtbl.add cache this x;
             x)
         (Hashtbl.find table this))
      |> Option.is_some
  in

  let table = Hashtbl.create 10000
  and paths = ref NodeSet.empty
  and cache = Hashtbl.create 10000 in

  eadges_in_table table edges;
  helper table cache paths source dest steps

let main =
  assert (valid_path 3 [ [ 0; 1 ]; [ 1; 2 ]; [ 2; 0 ]; [ 0; 2 ] ] 0 2);
  assert (
    not (valid_path 6 [ [ 0; 1 ]; [ 0; 2 ]; [ 3; 5 ]; [ 5; 4 ]; [ 4; 3 ] ] 0 5));
  assert (valid_path 1 [] 0 0);
  assert (
    not
      (valid_path 50
         [
           [ 31; 5 ];
           [ 10; 46 ];
           [ 19; 31 ];
           [ 5; 1 ];
           [ 31; 28 ];
           [ 28; 29 ];
           [ 8; 26 ];
           [ 13; 23 ];
           [ 16; 34 ];
           [ 30; 1 ];
           [ 16; 18 ];
           [ 33; 46 ];
           [ 27; 35 ];
           [ 2; 25 ];
           [ 49; 33 ];
           [ 44; 19 ];
           [ 22; 26 ];
           [ 30; 13 ];
           [ 27; 12 ];
           [ 8; 16 ];
           [ 42; 13 ];
           [ 18; 3 ];
           [ 21; 20 ];
           [ 2; 17 ];
           [ 5; 48 ];
           [ 41; 37 ];
           [ 39; 37 ];
           [ 2; 11 ];
           [ 20; 26 ];
           [ 19; 43 ];
           [ 45; 7 ];
           [ 0; 21 ];
           [ 44; 23 ];
           [ 2; 39 ];
           [ 27; 36 ];
           [ 41; 48 ];
           [ 17; 42 ];
           [ 40; 32 ];
           [ 2; 28 ];
           [ 35; 38 ];
           [ 3; 9 ];
           [ 41; 30 ];
           [ 5; 11 ];
           [ 24; 22 ];
           [ 39; 5 ];
           [ 40; 31 ];
           [ 18; 35 ];
           [ 23; 39 ];
           [ 20; 24 ];
           [ 45; 1 ];
         ]
         29 46))
